### PR TITLE
Haskell.nix

### PR DESCRIPTION
**Pull request by o1lo01ol1o**
add haskell.nix infrastructure

### CODE REVIEW

**Comment by ocharles** [comment]
Awesome! I wanted to get this ASAP so we can get CI set up. Will take a look tomorrow, thanks!

File: default.nix [hunk 1]
```diff
@@ -0,0 +1,34 @@
+{ # Fetch the latest haskell.nix and import its default.nix
+  haskellNix ? import (builtins.fetchTarball "https://github.com/input-output-hk/haskell.nix/archive/master.tar.gz") {}
```
Comment (ocharles) [inline default.nix:2]: Let use [Niv](https://github.com/nmattia/niv) and pin a specific revision.

File: default.nix [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+# haskell.nix provides access to the nixpkgs pins which are used by our CI,
+# hence you will be more likely to get cache hits when using these.
+# But you can also just use your own, e.g. '<nixpkgs>'.
+, nixpkgsSrc ? haskellNix.sources.nixpkgs-2003
```
Comment (ocharles) [inline default.nix:7]:
```suggestion
, nixpkgsSrc ? haskellNix.sources.nixpkgs-2009
```

File: default.nix [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+, nixpkgsArgs ? haskellNix.nixpkgsArgs
+, compiler-nix-name ? "ghc884"
+}: 
+let
+  # Define custom nixpkgs overlay
+  overlays = haskellNix.overlays ++ [
+    (self: super: {
+        sdl2 = super.haskellPackages.sdl2;
+    })
+  ];
```
Comment (ocharles) [inline default.nix:20]: This doesn't look right. I guess Haskell.nix tries to build the Haskell SDL2 library and complains about the `sdl2` `pkg-config` dependency? If so, we need to send a PR to Haskell.nix to solve this. Happy to look into this if you show me the error you get if you remove this overlay.

File: default.nix [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  src = pkgs.haskell-nix.haskellLib.cleanGit {
+    name = "dear-imgui";
+    src = ./.;
+  };
+  # modules = [{
+  #   # Replace `extra-libraries` dependencies
+  #   packages.GL.components.library.libs = pkgs.lib.mkForce (with pkgs;
+  #       [ darwin.apple_sdk.frameworks.OpenGL ]);
+  # }];
+
```
Comment (ocharles) [inline default.nix:33]: Can this go? I don't like to commit commented out code.

File: hs-dear-imgui.cabal [hunk 5]
```diff
@@ -1,5 +1,5 @@
 cabal-version: 3.0
-name: dear-imgui
+name: hs-dear-imgui
```
Comment (ocharles) [inline hs-dear-imgui.cabal:2]: I'd rather keep this library named `dear-imgui`, it's implicit that it's Haskell.

File: src/DearImGui.hs [hunk 6]
```diff
@@ -293,7 +293,7 @@ sameLine = liftIO do
 text :: MonadIO m => String -> m ()
 text t = liftIO do
   withCString t \textPtr ->
-    [C.exp| void { Text($(char* textPtr)) } |]
+    [C.exp| void { Text("%s", $(char* textPtr)) } |]
```
Comment (ocharles) [inline src/DearImGui.hs:296]: Is this change necessary? Maybe we can do it in a separate PR?

File: src/DearImGui/Context.hs [hunk 7]
```diff
@@ -25,7 +25,7 @@ instance Storable ImVec4 where
     poke (castPtr ptr `plusPtr` (sizeOf x * 3)) w
 
   peek ptr = do
-    x <- peek (castPtr ptr `plusPtr`               )
+    x <- peek (castPtr ptr                         )
```
Comment (ocharles) [inline src/DearImGui/Context.hs:28]: Whoops! I've fixed this in `master` now :smile:

File: default.nix (hunk 1 above)
Comment (o1lo01ol1o) [reply default.nix:2]: will do!

File: default.nix (hunk 4 above)
Comment (o1lo01ol1o) [reply default.nix:33]: Yes, of course, I'm going to clean and rebase when it's all working.

File: src/DearImGui.hs (hunk 6 above)
Comment (o1lo01ol1o) [reply src/DearImGui.hs:296]: This triggers a security error that prevents me from compiling on darwin, so, for the moment, yes it's necessary.  I can revert before merging if you like.

File: default.nix (hunk 3 above)
Comment (o1lo01ol1o) [reply default.nix:20]: Yes: 

```
error: The Nixpkgs package set does not contain the package: sdl2 (system dependency).
You may need to augment the system package mapping in haskell.nix so that it can be found.
```
per [`haskell.nixlib/pkgconf-nixpkgs-map.nix`](https://github.com/input-output-hk/haskell.nix/blob/36a85c8b3483b5a8fbca5ccabd0f71b08b2d9623/lib/pkgconf-nixpkgs-map.nix), `sdl2` is mapped to `SDL2` not `haskellPackages.sdl2`

Comment (ocharles) [reply default.nix:20]: Hmm, but that's correct - this is about the C library which _is_ `pkgs.SDL2`. Can I see the full build log up to failure? Is it failing to build `dear-imgui` or the `sdl2` dependency?

File: src/DearImGui.hs (hunk 6 above)
Comment (ocharles) [reply src/DearImGui.hs:296]: Oh, wow! Sounds like we should bring this fix in, but could we do it in a separate PR?

File: cabal.project [hunk 8]
```diff
@@ -0,0 +1,3 @@
+packages:
+    *.cabal
+
```
Comment (ocharles) [inline cabal.project:3]: I just realised we should probably enable all flags here, too.

File: hs-dear-imgui.cabal (hunk 5 above)
Comment (o1lo01ol1o) [reply hs-dear-imgui.cabal]: my fault, there was some confusion between haskell.nix, cabal, myself, and what "name"s were being used where. fixed now.

Comment (o1lo01ol1o) [reply hs-dear-imgui.cabal]: actually not.  It seems like haskell.nix expects some sort of parity between package names, cabal file names, and haskell.nix "name" identifiers.  after dropping `hs-` `nix-shell` fails with: 
```
error: The option `packages.dear-imgui.package.identifier.version' is used but not defined.
```
Do you know where the mixup might be?

File: src/DearImGui.hs (hunk 6 above)
Comment (o1lo01ol1o) [reply src/DearImGui.hs:296]: done in #6

File: default.nix (hunk 3 above)
Comment (o1lo01ol1o) [reply default.nix:20]:
```
% nix-build -A dear-imgui.components.library --show-trace
warning: Nix search path entry '/nix/var/nix/profiles/per-user/root/channels' does not exist, ignoring
trace: No index state specified for dear-imgui, using the latest index state that we know about (2021-01-25T00:00:00Z)!
error: while evaluating the attribute '__impureHostDeps' of the derivation 'dear-imgui-lib-dear-imgui-1.0.0' at /nix/store/9wkvnaxphqx96pq7ga6rbf1klxfp9xvq-source/pkgs/stdenv/generic/make-derivation.nix:192:11:
while evaluating anonymous function at /nix/store/9wkvnaxphqx96pq7ga6rbf1klxfp9xvq-source/pkgs/stdenv/generic/make-derivation.nix:175:36, called from /nix/store/9wkvnaxphqx96pq7ga6rbf1klxfp9xvq-source/pkgs/stdenv/generic/make-derivation.nix:175:21:
while evaluating 'chooseDevOutputs' at /nix/store/9wkvnaxphqx96pq7ga6rbf1klxfp9xvq-source/lib/attrsets.nix:475:22, called from undefined position:
while evaluating the attribute 'libs' at undefined position:
while evaluating 'g' at /nix/store/9wkvnaxphqx96pq7ga6rbf1klxfp9xvq-source/lib/attrsets.nix:276:19, called from undefined position:
while evaluating anonymous function at /nix/store/9wkvnaxphqx96pq7ga6rbf1klxfp9xvq-source/lib/modules.nix:98:72, called from /nix/store/9wkvnaxphqx96pq7ga6rbf1klxfp9xvq-source/lib/attrsets.nix:279:20:
while evaluating the attribute 'value' at /nix/store/9wkvnaxphqx96pq7ga6rbf1klxfp9xvq-source/lib/modules.nix:465:9:
while evaluating the option `packages.dear-imgui.components.library.libs':
while evaluating the attribute 'mergedValue' at /nix/store/9wkvnaxphqx96pq7ga6rbf1klxfp9xvq-source/lib/modules.nix:497:5:
while evaluating 'merge' at /nix/store/1f6fbxz6x90k90qfkwjmgrpivdz75w3l-source/modules/plan.nix:39:18, called from /nix/store/9wkvnaxphqx96pq7ga6rbf1klxfp9xvq-source/lib/modules.nix:499:59:
while evaluating anonymous function at /nix/store/1f6fbxz6x90k90qfkwjmgrpivdz75w3l-source/modules/plan.nix:40:33, called from undefined position:
while evaluating anonymous function at /nix/store/9wkvnaxphqx96pq7ga6rbf1klxfp9xvq-source/lib/lists.nix:116:29, called from undefined position:
while evaluating anonymous function at /nix/store/1f6fbxz6x90k90qfkwjmgrpivdz75w3l-source/modules/plan.nix:42:21, called from /nix/store/9wkvnaxphqx96pq7ga6rbf1klxfp9xvq-source/lib/lists.nix:116:32:
while evaluating the attribute 'optionalValue' at /nix/store/9wkvnaxphqx96pq7ga6rbf1klxfp9xvq-source/lib/modules.nix:509:5:
while evaluating the attribute 'values' at /nix/store/9wkvnaxphqx96pq7ga6rbf1klxfp9xvq-source/lib/modules.nix:491:9:
while evaluating the attribute 'values' at /nix/store/9wkvnaxphqx96pq7ga6rbf1klxfp9xvq-source/lib/modules.nix:590:7:
while evaluating anonymous function at /nix/store/9wkvnaxphqx96pq7ga6rbf1klxfp9xvq-source/lib/modules.nix:477:28, called from /nix/store/9wkvnaxphqx96pq7ga6rbf1klxfp9xvq-source/lib/modules.nix:477:17:
while evaluating definitions from `/nix/store/1f6fbxz6x90k90qfkwjmgrpivdz75w3l-source/modules/plan.nix':
while evaluating 'dischargeProperties' at /nix/store/9wkvnaxphqx96pq7ga6rbf1klxfp9xvq-source/lib/modules.nix:549:25, called from /nix/store/9wkvnaxphqx96pq7ga6rbf1klxfp9xvq-source/lib/modules.nix:478:137:
while evaluating the attribute 'value' at /nix/store/1f6fbxz6x90k90qfkwjmgrpivdz75w3l-source/modules/plan.nix:46:38:
while evaluating 'sysDepError' at /nix/store/1f6fbxz6x90k90qfkwjmgrpivdz75w3l-source/modules/error-handler.nix:13:19, called from /nix/store/7mcaxsq6sk28fi53gwdxygjp669wvgq4-dear-imgui-plan-to-nix-pkgs/.plan.nix/dear-imgui.nix:48:126:
The Nixpkgs package set does not contain the package: sdl2 (system dependency).
You may need to augment the system package mapping in haskell.nix so that it can be found
```

Comment (o1lo01ol1o) [reply default.nix:20]: on darwin, doesn't this condition say that `sdl2` should be the haskell package?

```
if os(windows) || os(darwin)
      extra-libraries:
        sdl2
```

Comment (ocharles) [reply default.nix:20]: No, `extra-libraries: sdl2` means that for `cabal build` to work `-lsdl2` needs to work. So this is about the libsdl dependency, not a Haskell dependency. We need this because we'll compile `imgui/backends/imgui_impl_sdl.cpp`, and that needs `-lsdl2`.

File: dear-imgui.cabal [hunk 9]
```diff
@@ -56,6 +56,8 @@ library
     if os(windows)
       extra-libraries:
         opengl32
+    if os(darwin)
+      frameworks: OpenGL
```
Comment (ocharles) [inline dear-imgui.cabal:60]: Do you mind submitting this in a separate PR? It's not quite related to haskell.nix stuff, and I like to keep PRs specific.

File: default.nix [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+, nixpkgsSrc ? haskellNix.sources.nixpkgs-2009
+
+  # haskell.nix provides some arguments to be passed to nixpkgs, including some
+  # patches and also the haskell.nix functionality itself as an overlay.
+, nixpkgsArgs ? haskellNix.nixpkgsArgs, compiler-nix-name ? "ghc884" }:
+let
+  pkgs' = import <nixpkgs> { };
+  # Define custom nixpkgs overlay
+  darwinOverlays = if pkgs'.stdenv.isDarwin then
+    [ (self: super: { sdl2 = super.haskellPackages.sdl2; }) ]
```
Comment (ocharles) [inline default.nix:16]: This still doesn't look right. This is saying that `pkgs.sdl2` - the `libsdl2` C library - is the Haskell package `sdl2`. These aren't the same thing. Are you sure you need this?

File: shell.nix [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+    ];
+
+    # Builds a Hoogle documentation index of all dependencies,
+    # and provides a "hoogle" command to search the index.
+    withHoogle = true;
+
+    # You might want some extra tools in the shell (optional).
+
+    # Some common tools can be added with the `tools` argument
+    tools = { cabal = "3.2.0.0"; hlint = "2.2.11"; };
```
Comment (ocharles) [inline shell.nix:17]: We're not using `hlint` atm, so let's drop this.

```suggestion
    tools = { cabal = "3.2.0.0"; };
```

File: shell.nix [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    # See overlays/tools.nix for more details
```
Comment (ocharles) [inline shell.nix:18]: This comment is a bit misleading as we don't have this file. I think it's referring to `haskell.nix/overlays/tools.nix`, but right now it's confusing - let's drop it:

```suggestion
```

File: shell.nix [hunk 13]
```diff
@@ -0,0 +1,23 @@
+let
+  hsPkgs = import ./default.nix {};
+in
+  hsPkgs.shellFor {
+    # Include only the *local* packages of your project.
+    packages = ps: with ps; [
+      dear-imgui
+    ];
+
```
Comment (ocharles) [inline shell.nix:9]: Haskell.nix should do this by default now, so we don't need to be explicit.

```suggestion
```

**Comment by TristanCacqueray** [comment]
Is there a reason to not use `nixpkgs.haskellPackages`? For what it worth, I've been able to run the example using this shell: `nix-shell --pure -p "haskellPackages.ghcWithPackages (p: [p.base p.sdl2 p.gl p.containers p.inline-c p.inline-c-cpp p.StateVar p.managed p.cabal-install ])" -p libGL -p SDL2 -p pkgconfig -p gcc`

**Comment by ocharles** [comment]
@TristanCacqueray I'm in favour of using Haskell.nix - I find it a much more reasonable way to do things (e.g., letting Cabal do the solving, rather than using LTS or some arbitrary package set). This PR does work, but we just had to do some tweaking to get it to work on Darwin.

Btw, you might want to come join us at `#dear-imgui.hs:ocharles.org.uk` on Matrix if you want to see the "live stream" :)

File: default.nix [hunk 14]
```diff
@@ -0,0 +1,21 @@
+{ # Fetch the latest haskell.nix and import its default.nix
```
Comment (ocharles) [inline default.nix:1]:
```suggestion
{ # Fetch haskell.nix and import its default.nix
```

File: default.nix [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+haskellNix ? (import (import ./nix/sources.nix)."haskell.nix" { })
+
+# haskell.nix provides access to the nixpkgs pins which are used by our CI,
+# hence you will be more likely to get cache hits when using these.
+# But you can also just use your own, e.g. '<nixpkgs>'.
+, nixpkgsSrc ? haskellNix.sources.nixpkgs-2009
+
+  # haskell.nix provides some arguments to be passed to nixpkgs, including some
+  # patches and also the haskell.nix functionality itself as an overlay.
+, nixpkgsArgs ? haskellNix.nixpkgsArgs, compiler-nix-name ? "ghc884" }:
```
Comment (ocharles) [inline default.nix:11]:
```suggestion
, nixpkgsArgs ? haskellNix.nixpkgsArgs
, compiler-nix-name ? "ghc884"
}:
```

**Comment by TristanCacqueray** [comment]
Alright, thank you for the explanation.
